include/libcyphal/application/registry/registry_impl.hpp

```diff
@@ -0,0 +1,124 @@
+/// @copyright
+/// Copyright (C) OpenCyphal Development Team  <opencyphal.org>
+/// Copyright Amazon.com Inc. or its affiliates.
+/// SPDX-License-Identifier: MIT
+
+#ifndef LIBCYPHAL_APPLICATION_REGISTRY_IMPL_HPP_INCLUDED
+#define LIBCYPHAL_APPLICATION_REGISTRY_IMPL_HPP_INCLUDED
+
+#include "libcyphal/common/cavl/cavl.hpp"
+#include "registry.hpp"
+
+#include <cetl/cetl.hpp>
+#include <cetl/pf17/cetlpf.hpp>
+
+namespace libcyphal
+{
+namespace application
+{
+namespace registry
+{
+
+/// Defines the registry implementation.
+///
+class Registry final : public IIntrospectableRegistry
+{
+public:
+    /// Constructs a new registry.
+    ///
+    /// @param memory The memory resource to use for variable size-d register values.
+    ///
+    explicit Registry(cetl::pmr::memory_resource& memory)
+        : memory_{memory}
+    {
+    }
+    ~Registry() = default;
+
+    Registry(Registry&&)                 = delete;
+    Registry(const Registry&)            = delete;
+    Registry& operator=(Registry&&)      = delete;
+    Registry& operator=(const Registry&) = delete;
+
+    cetl::pmr::memory_resource& memory() const noexcept
+    {
+        return memory_;
+    }
+
+    // MARK: - IRegistry
+
+    cetl::optional<IRegister::ValueAndFlags> get(const IRegister::Name name) const override
+    {
+        if (const auto* const reg = findRegisterBy(name))
+        {
+            return reg->get();
+        }
+        return cetl::nullopt;
+    }
+
+    cetl::optional<SetError> set(const IRegister::Name name, const Value& new_value) override
+    {
+        if (auto* const reg = findRegisterBy(name))
+        {
+            return reg->set(new_value);
+        }
+        return SetError::Existence;
+    }
+
+    // MARK: - IIntrospectableRegistry
+
+    std::size_t size() const override
+    {
+        return registers_tree_.size();
+    }
+
+    IRegister::Name index(const std::size_t index) const override
+    {
+        if (const auto* const reg = registers_tree_[index])
+        {
+            return reg->getName();
+        }
+        return IRegister::Name{};
+    }
+
+    bool append(IRegister& reg) override
```

Comment:
Why can't we make factory methods here that create a register instance and then add it in one operation?

Reply:
Done - see `Registry::route`, `Registry::expose` and `Registry::exposeParam` methods.
Please note that I intentionally left my original `makeRegister` methods as is b/c:
- `route` overload methods use them
- user could use them if he wants:
  - first **infallibly** create registers (f.e. as members during constructor),
  - and only later link them to the registry (which might fail due to name duplicates).

Added more docs mentioning all possible factory alternatives.